### PR TITLE
Add documentation for DFRobot mmWave Radar

### PR DESCRIPTION
**Pull request by niklasweber**
## Description:

Add documentation for new DFRobot mmWave Radar component

**Related issue (if applicable):** -

**Pull request in [esphome](https://github.com/esphome/esphome) with YAML changes (if applicable):** esphome/esphome#4203

## Checklist:

  - [x] I am merging into `next` because this is new documentation that has a matching pull-request in [esphome](https://github.com/esphome/esphome) as linked above.  
    or
  - [ ] I am merging into `current` because this is a fix, change and/or adjustment in the current documentation and is not for a new component or feature.

  - [x] Link added in `/index.rst` when creating new documents for new components or cookbook.

### CODE REVIEW

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 1]
```diff
@@ -0,0 +1,282 @@
+DFRobot mmWave Radar
+===============================
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]:
```suggestion
DFRobot mmWave Radar
====================
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+
+    switch:
+      # Switch to turn on/off detection
+      - platform: dfrobot_mmwave_radar
+        name: Mmwave Active
+
+.. _dfr_mmwave-component:
+
+Component
+-------------
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]:
```suggestion
Component
---------
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    dfrobot_mmwave_radar:
+      - id: mmwave_1
+        uart_id: uart_1
+      - id: mmwave_2
+        uart_id: uart_2
+      ...
+
+**Configuration variables**:
+
+- **id** (*Optional*, :ref:`config-id`): Manually specify the ID used for code generation. Necessary if you want to define multiple mmwave radars.
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]:
```suggestion
- **id** (*Optional*, :ref:`config-id`): Manually specify the ID used for code generation. Necessary if you want to define multiple instances of this component.
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+- **uart_id** (*Optional*, :ref:`config-id`): Manually specify the ID of the UART Component if you want to use multiple UART buses.
+
+Binary Sensor
+-------------
+
+via GPIO
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]:
```suggestion
Via GPIO
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+    binary_sensor:
+      - platform: gpio
+        name: Mmwave Detected via GPIO
+        device_class: motion
+        pin:
+          number: GPIO15
+          mode: INPUT_PULLDOWN
+
+via UART
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]:
```suggestion
Via UART
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+********
+
+If you want, you can use the sensor only using a uart connection. It might be an advantage to be able to save an extra pin. However, the CPU
+needs to be running in order to process incoming sensor data. You cannot wake up the processor by using uart only.
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]: Rephrasing this a bit.
```suggestion
If necessary, you can use the sensor by way of the serial connection alone; this reduces the number of pins required. However, the CPU
needs to be running in order to process incoming sensor data. The UART cannot wake processor.
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+
+    binary_sensor:
+      - platform: dfrobot_mmwave_radar
+        name: Mmwave Detected via UART
+
+Configuration variables:
+
+- **id** (*Optional*, :ref:`config-id`): Manually specify the ID used for code generation.
+- **name** (*Optional*, string): The name for the binary sensor.
+- **dfrobot_mmwave_radar_id** (*Optional*, :ref:`config-id`): The ID of the DFRobot mmWave component defined above. Useful for multiple mmWave sensors.
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]:
```suggestion
- **dfrobot_mmwave_radar_id** (*Optional*, :ref:`config-id`): The ID of the DFRobot mmWave component defined above. Useful when multiple instances of this component are defined.
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+- All other options from :ref:`Binary Sensor <config-binary_sensor>`.
+
+
+Switch
+-------------
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]:
```suggestion
Switch
------
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+
+You can use the switch to get an indication whether the sensor is currently active or not and you can use it to turn the sensor on and off.
+Don't confuse it with the detection state. When turned off, you won't get detected when in front of the sensor.
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]: Rephrasing this a bit to be more concise
```suggestion
The switch is used to enable and/or disable detection. *Note that this does not indicate the state of detection.*
When the switch is turned off, you won't get detected when in front of the sensor.
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+
+    switch:
+      - platform: dfrobot_mmwave_radar
+        name: Mmwave Active
+        dfrobot_mmwave_radar_id: mmwave
+
+Configuration variables:
+
+- **id** (*Optional*, :ref:`config-id`): Manually specify the ID used for code generation.
+- **dfrobot_mmwave_radar_id** (*Optional*, :ref:`config-id`): The ID of the DFRobot mmWave component defined above. Useful for multiple mmWave sensors.
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]:
```suggestion
- **dfrobot_mmwave_radar_id** (*Optional*, :ref:`config-id`): The ID of the DFRobot mmWave component defined above. Useful when multiple instances of this component are defined.
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+
+``dfrobot_mmwave_radar.settings`` Action
+----------------------------------------
+
+.. warning::
+
+    As it seems every time you save the configuration of the mmwave radar, it is written to its internal flash or EEPROM
+    which can be susceptible to damage when writing a lot. Do not change setting on every boot or several times per second,
+    otherwise the mmwave radar might suffer damage. Instead, find the correct settings during setup phase and do not regularly change
+    settings later on.
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]: Rephrasing this a bit:
```suggestion
    Each change to the configuration of the mmWave radar triggers a write to its internal flash/EEPROM.
    Write cycles to this memory are limited, so do not change settings frequently or this memory may wear out.
    Determine the appropriate settings for your device and avoid changing them unless absolutely necessary.
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+
+The DFRobot mmWave Radar has several settings which can be changed. These settings are saved in non-volatile memory and do not need to be set on each boot.
+
+All options of the settings action are optional and only the specified options are changed. Unspecified parameters are untouched (not set to a default value).
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]:
```suggestion
The settings action allows changing of any number of the mmWave radar's internal parameters/settings.
With this action, any unspecified parameters will remain unchanged.
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+          presence_via_uart: true
+          sensitivity: 7
+
+
+Configuration options:
+
+- **id** (*Optional*, :ref:`config-id`): Manually specify the ID of the mmwave sensor which settings should be changed. If only one radar is defined, this is optional.
+
+- **factory_reset** (*Optional*, boolean): If set to true, a factory reset of the sensor will be performed (before changing other options if present). Ignored if not set or set to false.
+  A lambda function can be used to set the value. Return true or false.
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]:
```suggestion
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+  Value is tempatable: Return true or false. Returning -1 keeps the value unchanged.
+
+- **sensitivity** (*Optional*, int): Set the sensitivity of the sensor. Ranges from 0 to 9. Value is tempatable: Return 0-9. Returning -1 keeps the value unchanged.
+
+``dfrobot_mmwave_radar.start`` Action
+----------------------------------------
+
+.. note::
+
+    It seems the DFRobot mmWave Radar has a bug in it's firmware. If the uart tx line (rx on the sensor) is held low (e.g. during serial flashing),
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]:
```suggestion
    It seems the DFRobot mmWave Radar has a bug in its firmware. If the UART TX line (RX on the sensor) is held low (e.g. during serial flashing),
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+
+    on_...:
+      dfrobot_mmwave_radar.start
+
+Configuration options:
+
+- **id** (*Optional*, :ref:`config-id`): Manually specify the ID of the mmwave component if you have multiple ones.
+
+``dfrobot_mmwave_radar.stop`` Action
+----------------------------------------
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]:
```suggestion
``dfrobot_mmwave_radar.stop`` Action
------------------------------------
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 16]
```diff
@@ -0,0 +1,282 @@
...
+
+    on_...:
+      dfrobot_mmwave_radar.stop
+
+Configuration options:
+
+- **id** (*Optional*, :ref:`config-id`): Manually specify the ID of the mmwave component if you have multiple ones.
+
+``dfrobot_mmwave_radar.reset`` Action
+----------------------------------------
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]:
```suggestion
- **id** (*Optional*, :ref:`config-id`): Manually specify the ID of the mmWave component. Useful when multiple instances of this component are defined.

``dfrobot_mmwave_radar.reset`` Action
-------------------------------------
```

File: components/binary_sensor/dfrobot_mmwave.rst [hunk 17]
```diff
@@ -0,0 +1,282 @@
...
+Restart the sensor.
+
+.. code-block:: yaml
+
+    on_...:
+      dfrobot_mmwave_radar.reset
+
+Configuration options:
+
+- **id** (*Optional*, :ref:`config-id`): Manually specify the ID of the mmwave component if you have multiple ones.
```
Comment (kbx81) [inline components/binary_sensor/dfrobot_mmwave.rst]:
```suggestion
- **id** (*Optional*, :ref:`config-id`): Manually specify the ID of the mmWave component. Useful when multiple instances of this component are defined.
```

**Review by kbx81 — DISMISSED** [review]
Just various nitpicks/rephrasing 😇

File: components/binary_sensor/dfrobot_mmwave.rst
Comment (jesserockz) [inline components/binary_sensor/dfrobot_mmwave.rst:1]: This file should be renamed to `components/dfrobot_sen0395.rst`

Comment (kbx81) [reply components/binary_sensor/dfrobot_mmwave.rst:1]: Done

File: components/binary_sensor/dfrobot_sen0395.rst [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+
+    DFRobot mmWave Radar / presence detection sensor
+
+.. code-block:: yaml
+
+    # Example configuration entry
+    uart:
+      tx_pin: GPIO13
+      rx_pin: GPIO14
+      baud_rate: 115200
```
Comment (jesserockz) [inline components/binary_sensor/dfrobot_sen0395.rst]: Its preferred to not include this as it forces the user to go to the uart docs and read those.
```suggestion
```

File: components/binary_sensor/dfrobot_sen0395.rst [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+      # Information coming from uart bus
+      - platform: dfrobot_sen0395
+        name: Mmwave Detected via UART
+
+      # Information coming from dedicated gpio (IO2)
+      - platform: gpio
+        name: Mmwave Detected via GPIO
+        device_class: motion
+        pin:
+          number: GPIO15
```
Comment (jesserockz) [inline components/binary_sensor/dfrobot_sen0395.rst]: One thing I started doing for new docs with gpio pins is setting the value to `REPLACEME`. The ESPHome config loader handles these and throws an error forcing the end user to put their correct pin in.

```suggestion
          number: REPLACEME
```

File: components/binary_sensor/dfrobot_sen0395.rst [hunk 20]
```diff
@@ -0,0 +1,274 @@
...
+``dfrobot_sen0395:`` and ``uart:`` components are not necessary.
+
+.. code-block:: yaml
+
+    binary_sensor:
+      - platform: gpio
+        name: Presence Detected via GPIO
+        device_class: motion
+        pin:
+          number: GPIO15
```
Comment (jesserockz) [inline components/binary_sensor/dfrobot_sen0395.rst]:
```suggestion
          number: REPLACEME
```

File: components/binary_sensor/dfrobot_sen0395.rst [hunk 21]
```diff
@@ -0,0 +1,274 @@
...
+
+    binary_sensor:
+      - platform: dfrobot_sen0395
+        name: Presence Detected via UART
+
+Configuration variables:
+************************
+
+- **id** (*Optional*, :ref:`config-id`): Manually specify the ID used for code generation.
+- **name** (*Optional*, string): The name for the binary sensor.
```
Comment (jesserockz) [inline components/binary_sensor/dfrobot_sen0395.rst]: Similar to the above comment about `uart` in examples, don't add `id` and `name` to config variables list as users can see them in the "binary sensor" core config page
```suggestion
```

File: components/dfrobot_sen0395.rst [hunk 22]
```diff
@@ -0,0 +1,267 @@
...
+
+``dfrobot_sen0395.reset`` Action
+********************************
+
+Restart the sensor.
+
+.. code-block:: yaml
+
+    on_...:
+      dfrobot_sen0395.reset
```
Comment (jesserockz) [inline components/dfrobot_sen0395.rst]: Can't forget the colons
```suggestion
      dfrobot_sen0395.reset:
```

**Review by jesserockz — APPROVED** [review]

**Pull request merged**